flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/source/enumerator/CassandraSplitEnumerator.java

```diff
@@ -0,0 +1,156 @@
+public final class CassandraSplitEnumerator
+        implements SplitEnumerator<CassandraSplit, CassandraEnumeratorState> {
+    private static final Logger LOG = LoggerFactory.getLogger(CassandraSplitEnumerator.class);
+    private static final String MURMUR3PARTITIONER = "org.apache.cassandra.dht.Murmur3Partitioner";
+
+    private final SplitEnumeratorContext<CassandraSplit> enumeratorContext;
+    private final CassandraEnumeratorState state;
+    private final Cluster cluster;
+
+    public CassandraSplitEnumerator(
+            SplitEnumeratorContext<CassandraSplit> enumeratorContext,
+            CassandraEnumeratorState state,
+            ClusterBuilder clusterBuilder) {
+        this.enumeratorContext = enumeratorContext;
+        this.state = state == null ? new CassandraEnumeratorState() : state /* snapshot restore*/;
+        this.cluster = clusterBuilder.getCluster();
+    }
+
+    @Override
+    public void handleSplitRequest(int subtaskId, @Nullable String requesterHostname) {
+        assignUnprocessedSplitsToReader(subtaskId);
+    }
+
+    @Override
+    public void start() {
+        // discover the splits and update unprocessed splits and then assign them.
+        // There is only an initial splits discovery, no periodic discovery.
+        enumeratorContext.callAsync(
+                this::discoverSplits,
+                (splits, throwable) -> {
+                    LOG.info("Add {} splits to CassandraSplitEnumerator.", splits.size());
+                    state.addNewSplits(splits, enumeratorContext.currentParallelism());
+                });
+    }
+
+    private List<CassandraSplit> discoverSplits() {
+        final int numberOfSplits = enumeratorContext.currentParallelism();
+        final Metadata clusterMetadata = cluster.getMetadata();
+        final String partitioner = clusterMetadata.getPartitioner();
+        final SplitsGenerator splitsGenerator = new SplitsGenerator(partitioner);
+        if (MURMUR3PARTITIONER.equals(partitioner)) {
+            LOG.info("Murmur3Partitioner detected, splitting");
+            List<BigInteger> tokens =
+                    clusterMetadata.getTokenRanges().stream()
+                            .map(
+                                    tokenRange ->
+                                            new BigInteger(
+                                                    tokenRange.getEnd().getValue().toString()))
+                            .collect(Collectors.toList());
+            return splitsGenerator.generateSplits(numberOfSplits, tokens);
+        } else {
+            // Murmur3Partitioner is the default and recommended partitioner for Cassandra 1.2+
+            // see
+            // https://docs.datastax.com/en/cassandra-oss/3.x/cassandra/architecture/archPartitionerAbout.html
+            LOG.warn(
+                    "The current Cassandra partitioner is {}, only Murmur3Partitioner is supported "
```

Comment:
What about the random partitioner? Shouldn't it be possible to create multiple splits?

Reply:
See my general comments about Apache Beam Cassandra partitionner